### ISSUE TITLE
Wrapping lines of a function signature that uses the new python 3.12 generic type parameter syntax

### ISSUE THREAD

**Issue by Questionsdenombres** [psf/black#4071]
**Context:**
- When a function has many parameters, it's more readable when on multiple lines, enforcing black to wrap lines adding a comma after the last parameter.
- See https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#how-black-wraps-lines
```python
def func(a: int, b: int) -> int:
    return a + b 

def func(
    a: int,
    b: int,
) -> int:
    return a + b
```

**python 3.12 new feature:**
- https://docs.python.org/3/whatsnew/3.12.html#summary-release-highlights
- A generic function can now have type parameters used by the static type checker.
- See [PEP 695](https://docs.python.org/3/whatsnew/3.12.html#whatsnew312-pep695), [type](https://docs.python.org/3/reference/simple_stmts.html#type) parameter syntax and the type statement
- See https://peps.python.org/pep-0695/#summary-examples
```python
def func[T](a: T, b: T) -> T:
    return a
```

**Example in the current _Black_ style:**
Enforcing black to wrap lines gives the ugly:
```python
def func[
    T
](a: T, b: T,) -> T:
    return a
```

**Desired style:**
```python
def func[T](
    a: T,
    b: T,
) -> T:
    return a + b
```

**Additional context**
python==3.12
black==23.11.0
Windows==11
VScode==1.84.2
VScode extension! Black Formatter==v2023.6.0

**Credits:**
Thank you for your great tool.
Patrick, Paris, France.

**Comment by JelleZijlstra** [comment]
Yes, we should change this. PR welcome!

**Comment by JackismyShephard** [comment]
What is the ETA on this?

**Comment by hauntsaninja** [comment]
Are you interested in making a PR?

**Comment by JackismyShephard** [comment]
Well, sadly I don't quite have the time for that right now.

**Comment by AndreaCensi** [comment]
@JackismyShephard if you need to undo what `black` did, try `ruff`: it will do exactly what you want

**Comment by FlorinAsavoaie** [comment]
FYI: An attempt to fix this is found in #4272 but doesn't seem trivial.

**Comment by Feuermurmel** [comment]
Just converted a large-ish code-base to the new syntax and found an even more egregious example. 😅 I thought that this example could maybe serve as an additional test case because the return type is also wrapped:

```python
def something_something_function[
    T: Model
](param: list[int], other_param: type[T], *, some_other_param: bool = True) -> QuerySet[
    T
]:
    pass
```

Where I would prefer this instead. :)

```python
def something_something_function[T: Model](
    param: list[int], other_param: type[T], *, some_other_param: bool = True
) -> QuerySet[T]:
	pass
```

**Comment by Pedro-Muller29** [comment]
### Context

I am currently working on a PR for this, and I am facing some doubts on what should be the valid syntax on some trailing comma cases:

```python
def trailing_comma1[T=int,](a: str):
    pass

def trailing_comma2[T=int](a: str,):
    pass
```

### Current Behaviour

Right now the expected output in the `type_param_defaults.py` test case gives:

```python
def trailing_comma1[
    T = int,
](a: str):
    pass


def trailing_comma2[
    T = int
](a: str,):
    pass
```

### My Assessment

I believe the following adjustments should be made to align with valid and consistent formatting:

1. `trailing_comma1` should keep the same output.
2. `trailing_comma2` output should change to:
```python
def trailing_comma2[T = int](
    a: str,
):
    pass
```

**Am I correct in my assessment?**

**Comment by hauntsaninja** [comment]
Yes, that sounds right (assuming default settings, not `--skip-magic-trailing-comma`)
